platform/MCU/nRF52840/src/flash_mal.c

```diff
@@ -593,37 +642,52 @@ bool FLASH_UpdateModules(void (*flashModulesCallback)(bool isUpdating))
         const size_t magic_num_offs = module_offs + offsetof(platform_flash_modules_t, magicNumber);
         uint16_t magic_num = 0;
         if (dct_read_app_data_copy(magic_num_offs, &magic_num, sizeof(magic_num)) != 0) {
-            return false;
+            return FLASH_ACCESS_RESULT_ERROR;
         }
         if (magic_num == 0xabcd) {
             // Copy module info
             if (dct_read_app_data_copy(module_offs, &modules[module_count], sizeof(platform_flash_modules_t)) != 0) {
-                return false;
+                return FLASH_ACCESS_RESULT_ERROR;
             }
             // Mark slot as unused
             magic_num = 0xffff;
             if (dct_write_app_data(&magic_num, magic_num_offs, sizeof(magic_num)) != 0) {
-                return false;
+                return FLASH_ACCESS_RESULT_ERROR;
             }
             ++module_count;
         }
         module_offs += sizeof(platform_flash_modules_t);
     }
+
+    for (size_t i = 0; i < module_count; ++i) {
+        if (modules[i].module_function == MOD_FUNC_BOOTLOADER && i != (module_count - 1)) {
+            // Perform bootloader update last as it will cause an automatic reset
+            const platform_flash_modules_t mod = modules[i];
+            memmove(&modules[i], &modules[i + 1], sizeof(platform_flash_modules_t) * (module_count - 1));
+            memcpy(&modules[module_count - 1], &mod, sizeof(mod));
+        }
+    }
+
+    int result = FLASH_ACCESS_RESULT_OK;
     for (size_t i = 0; i < module_count; ++i) {
         const platform_flash_modules_t* module = &modules[i];
         // Turn On RGB_COLOR_MAGENTA toggling during flash updating
         if (flashModulesCallback) {
             flashModulesCallback(true);
         }
         // Copy memory from source to destination based on flash device id
-        FLASH_CopyMemory(module->sourceDeviceID, module->sourceAddress, module->destinationDeviceID,
+        int r = FLASH_CopyMemory(module->sourceDeviceID, module->sourceAddress, module->destinationDeviceID,
                 module->destinationAddress, module->length, module->module_function, module->flags);
         // Turn Off RGB_COLOR_MAGENTA toggling
         if (flashModulesCallback) {
             flashModulesCallback(false);
         }
+
+        if (r == FLASH_ACCESS_RESULT_RESET_PENDING) {
```

Comment:
The error handling seems a bit off here, even though practically the bootloader doesn't care if this function fails and `FLASH_ACCESS_RESULT_RESET_PENDING` can never be returned on this platform.

If we want this function to try to copy all the modules and ignore errors caused by individual modules (the current behavior here and in `develop`), let's

a) Return an error if copying has failed for some of the modules.

b) If an error occurred while copying one module and reset is pending to apply another, `FLASH_ACCESS_RESULT_RESET_PENDING` takes precedence